src/Raven.Server/Documents/ShardedHandlers/ShardedQueriesHandler.cs

```diff
@@ -0,0 +1,843 @@
+                    tasks.Add(task);
+                }
+
+                await Task.WhenAll(tasks);
+
+                foreach (var (_, cmd) in _commands)
+                {
+                    _result.ResultEtag = Hashing.Combine(_result.ResultEtag, cmd.Result.ResultEtag);
+                    _result.IsMapReduce = cmd.Result.IsMapReduce;
+                }
+            }
+            
+            public ValueTask HandleIncludes()
+            {
+                HashSet<string> missing = null;
+                foreach (var (_, cmd) in _commands)
+                {
+                    if (cmd.Result.Includes == null || cmd.Result.Includes.Count == 0)
+                        continue;
+                    _result.Includes ??= new List<BlittableJsonReaderObject>();
+                    foreach (var id in cmd.Result.Includes.GetPropertyNames())
+                    {
+                        if (cmd.Result.Includes.TryGet(id, out BlittableJsonReaderObject include) && include != null)
+                        {
+                            _result.Includes.Add(include);
+                        }
+                        else
+                        {
+                            (missing ??= new HashSet<string>()).Add(id);
+                        }
+                    }
+                }
+
+                if (missing == null || missing.Count == 0)
+                {
+                    return ValueTask.CompletedTask;
+                }
+                return HandleIncludesAsync(missing);
+            }
+
+            private async ValueTask HandleIncludesAsync(HashSet<string> missing)
+            {
+                var includeCommands = new Dictionary<int, ShardedQueryCommand>();
+                try
+                {
+                    GenerateLoadByIdQueries(missing, includeCommands);
+                    var tasks = new List<Task>();
+                    foreach (var (shardIndex, cmd) in includeCommands)
+                    {
+                        tasks.Add(_parent.ShardedContext.RequestExecutors[shardIndex].ExecuteAsync(cmd, cmd.Context));
+                    }
+
+                    await Task.WhenAll(tasks);
+                    foreach (var (_, cmd) in includeCommands)
+                    {
+                        foreach (BlittableJsonReaderObject result in cmd.Result.Results)
+                        {
+                            _result.Includes.Add(result);
+                        }
+                    }
+                }
+                finally
+                {
+                    foreach (var (var, cmd) in includeCommands)
+                    {
+                        cmd.Dispose();
+                    }
+                }
+            }
+
+            public void OrderResults()
+            {
+                if (_query.Metadata.OrderBy?.Length > 0)
+                {
+                    _result.Results.Sort(this);
```

Comment:
the results from the shards are already ordered, so we can maybe improve performance by interleave sorting

Reply:
https://issues.hibernatingrhinos.com/issue/RavenDB-18167